frame/assets-freezer/src/lib.rs

```diff
@@ -0,0 +1,361 @@
+use frame_support::{ensure, dispatch::{DispatchError, DispatchResult}};
+use frame_support::traits::{
+	StoredMap, tokens::{
+		WithdrawConsequence, DepositConsequence, fungibles, fungibles::InspectHold, FrozenBalance,
+		WhenDust
+	}
+};
+use frame_system::Config as SystemConfig;
+
+pub use pallet::*;
+
+type BalanceOf<T> = <<T as Config>::Assets as fungibles::Inspect<<T as SystemConfig>::AccountId>>::Balance;
+type AssetIdOf<T> = <<T as Config>::Assets as fungibles::Inspect<<T as SystemConfig>::AccountId>>::AssetId;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+	use super::*;
+
+	/// The information concerning our freezing.
+	#[derive(Eq, PartialEq, Clone, Encode, Decode, RuntimeDebug, Default)]
+	pub struct FreezeData<Balance> {
+		/// The amount of funds that have been reserved. The actual amount of funds held in reserve
+		/// (and thus guaranteed of being unreserved) is this amount less `melted`.
+		///
+		/// If this `is_zero`, then the account may be deleted. If it is non-zero, then the assets
+		/// pallet will attempt to keep the account alive by retaining the `minimum_balance` *plus*
+		/// this number of funds in it.
+		pub(super) reserved: Balance,
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	/// The module configuration trait.
+	pub trait Config: frame_system::Config {
+		/// The overarching event type.
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		/// The fungibles trait impl whose assets this reserves.
+		type Assets: fungibles::InspectWithoutFreezer<Self::AccountId>;
+
+		/// Place to store the fast-access freeze data for the given asset/account.
+		type Store: StoredMap<(AssetIdOf<Self>, Self::AccountId), FreezeData<BalanceOf<Self>>>;
+	}
+
+	//
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	#[pallet::metadata(T::AccountId = "AccountId", BalanceOf<T> = "Balance", AssetIdOf<T> = "AssetId")]
+	pub enum Event<T: Config> {
+		/// An asset has been reserved.
+		/// \[asset, who, amount\]
+		Held(AssetIdOf<T>, T::AccountId, BalanceOf<T>),
+		/// An asset has been unreserved.
+		/// \[asset, who, amount\]
+		Released(AssetIdOf<T>, T::AccountId, BalanceOf<T>),
+	}
+
+	// No new errors
+	#[pallet::error]
+	pub enum Error<T> {}
+
+	// No hooks.
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {}
+
+	// No calls.
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {}
+}
+
+impl<T: Config> FrozenBalance<AssetIdOf<T>, T::AccountId, BalanceOf<T>> for Pallet<T> {
+	fn frozen_balance(id: AssetIdOf<T>, who: &T::AccountId) -> Option<BalanceOf<T>> {
+		let f = T::Store::get(&(id, who.clone()));
+		if f.reserved.is_zero() { None } else { Some(f.reserved) }
+	}
+}
+
+impl<T: Config> fungibles::Inspect<<T as SystemConfig>::AccountId> for Pallet<T> {
```

Comment:
DQ: why does this pallet itself need to implement this? Anywhere where this is used, a `T::Assets` is also present where it does implement `Fungibles`. Internally, `T::Assets::fn` and externally `Runtime::AssetFreezer::Assets::fn` can expose everything needed.

Reply:
ahh `T::Assets` is only `InspectWithoutFreezer`.